File: Problems/Python 3/328.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        evenHead = even = ListNode(None)
        oddHead = odd = ListNode(None)
        
        while head:
            odd.next = head
            odd = odd.next
            even.next = head.next
            even = even.next
            head = head.next.next if even else None
        
        odd.next = evenHead.next
        
        return oddHead.next

File: Problems/C/328.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* oddEvenList(struct ListNode* head){
    
    struct ListNode* start = malloc(sizeof(struct ListNode));
    struct ListNode *odd = start, *evenHead = start, *even = head;
    
    start->next = head;
    bool isOdd = true;
    
    while (even != NULL) {
        if (isOdd && odd->next != even)
        {
            evenHead->next = even->next;
            even->next = odd->next;
            odd->next = even;
            even = evenHead->next;
            odd = odd->next;
        } 
        else 
        {
            odd = isOdd ? odd->next : odd;
            evenHead = even;
            even = even->next;
        }
        
        isOdd = !isOdd;
    }
    
    return start->next;
}
